changes/changes.go

```diff
@@ -0,0 +1,612 @@
+		res.Summary.Delta = EmptyObject{}
+	}
+
+	err := svc.maybeLoadStats(ctx, leftDs)
+	if err != nil {
+		return nil, err
+	}
+	err = svc.maybeLoadStats(ctx, rightDs)
+	if err != nil {
+		return nil, err
+	}
+
+	leftStats, err := svc.parseStats(leftDs)
+	if err != nil {
+		return nil, err
+	}
+	rightStats, err := svc.parseStats(rightDs)
+	if err != nil {
+		return nil, err
+	}
+
+	res.Columns, err = svc.matchColumns(leftColCount, rightColCount, leftColItems, rightColItems, leftStats, rightStats)
+	if err != nil {
+		log.Debugf("failed to calculate stats change report: %s", err.Error())
+		return nil, qerr.New(err, "failed to calculate stats change report")
+	}
+
+	return res, nil
+}
+
+// matchColumns attempts to match up columns from the left and right side based on the column name
+// this is not ideal as datasets without a header have generic column names and in case of adding a column
+// before the end might shift the alignment and break comparison due to type differences of columns which
+// are not properly handled yet
+func (svc *Service) matchColumns(leftColCount, rightColCount int, leftColItems, rightColItems tabular.Columns, leftStats, rightStats []EmptyObject) ([]*ChangeReportDeltaComponent, error) {
+	maxColCount := leftColCount
+	if rightColCount > maxColCount {
+		maxColCount = rightColCount
+	}
+
+	columns := make([]*ChangeReportDeltaComponent, maxColCount)
+
+	type cIndex struct {
+		LPos int
+		RPos int
+	}
+
+	colIndex := map[string]*cIndex{}
+	for i := 0; i < maxColCount; i++ {
+		if i < leftColCount {
+			if c, ok := colIndex[leftColItems[i].Title]; ok && c != nil {
+				colIndex[leftColItems[i].Title].LPos = i
+			} else {
+				colIndex[leftColItems[i].Title] = &cIndex{
+					LPos: i,
+					RPos: -1,
+				}
+			}
+		}
+		if i < rightColCount {
+			if c, ok := colIndex[rightColItems[i].Title]; ok && c != nil {
+				colIndex[rightColItems[i].Title].RPos = i
+			} else {
+				colIndex[rightColItems[i].Title] = &cIndex{
+					LPos: -1,
+					RPos: i,
+				}
+			}
+		}
+	}
+
+	i := 0
+	for k, v := range colIndex {
+		columns[i] = &ChangeReportDeltaComponent{
+			Title: k,
+		}
+		var lCol, rCol *tabular.Column
+		if v.LPos >= 0 {
+			columns[i].Left = leftStats[v.LPos]
+			lCol = &leftColItems[v.LPos]
+		} else {
+			columns[i].Left = EmptyObject{}
+		}
+		if v.RPos >= 0 {
+			columns[i].Right = rightStats[v.RPos]
+			rCol = &rightColItems[v.RPos]
+		} else {
+			columns[i].Right = EmptyObject{}
+		}
+		deltaCol, err := svc.columnStatsDelta(columns[i].Left, columns[i].Right, lCol, rCol, v.LPos >= 0, v.RPos >= 0)
+		if err != nil {
+			log.Debugf("error calculating stats delta: %s", err.Error())
+			return nil, qerr.New(err, fmt.Sprintf("failed to calculate stats column delta for %q", columns[i].Title))
+		}
+		columns[i].Delta = deltaCol
+		i++
+	}
+
+	return columns, nil
+}
+
+func (svc *Service) columnStatsDelta(left, right interface{}, lCol, rCol *tabular.Column, hasLeft, hasRight bool) (map[string]interface{}, error) {
```

Comment:
This function is really dense. Wondering if there's some way to split it up into smaller pieces, or refactor or combine code.

Reply:
Most of them are :( Will try another pass to make things a bit leaner.